### PR TITLE
CP-29991: Host selection for multiple vGPU

### PR DESCRIPTION
**Pull request by liulinC**
Host selection for multiple vGPU

test done:
- manual dev test

### CODE REVIEW

**Comment by liulinC** [comment]
@robhoes @lindig  could you please help to review this PR?
Thanks for your nice help about this
Regards!
Lin

**Comment by liulinC** [comment]
@lindig @robhoes May I know is there any further concern about this PR?

File: ocaml/xapi/xapi_vm_helpers.ml [hunk 1]
```diff
@@ -722,16 +692,32 @@ let get_group_key ~__context ~vm =
...
-  let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu in
-  let gpu_group = Db.VGPU.get_GPU_group ~__context ~self:vgpu in
-  match
-    Xapi_gpu_group.get_remaining_capacity_internal ~__context
-      ~self:gpu_group ~vgpu_type
-  with
-  | Either.Left e -> raise e
-  | Either.Right _ -> ();
-    group_hosts_by_best_pgpu_in_group ~__context gpu_group vgpu_type
+(* Rank the provided hosts (visible_hosts) by the remainding capacity for the
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_helpers.ml]: Mostly grammar:
Rank the provided hosts by their remaining capacity for the given vGPU. The vGPU has an associated vGPU group. We get the rank preference from the vGPU group.

File: ocaml/xapi/xapi_vm_helpers.ml [hunk 2]
```diff
@@ -722,16 +692,32 @@ let get_group_key ~__context ~vm =
...
+ * give vGPU from the vGPU associated group
+ * The rank preference is get from the vgpu associated gpu_group
+ * vgpu: the vGPU used to rank the provided hosts
+ * visible_hosts: the hosts will be ranked*)
+let group_hosts_by_best_vgpu ~__context vgpu visible_hosts =
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_helpers.ml]: Is "group_hosts" the best phrase? Isn't this providing a ranking such that "rank_hosts" would be better? I would understand grouping as partitioning of hosts.

**Comment by liulinC** [comment]
@lindig Thanks for your comments, updated, May I ask your approve if no other comments?
Thanks in advance.

File: ocaml/xapi/xapi_vm_helpers.ml [hunk 3]
```diff
@@ -722,16 +692,32 @@ let get_group_key ~__context ~vm =
...
+      | `depth_first -> `ascending | `breadth_first -> `descending
+    in
+    Helpers.group_by ~ordering
+      (fun host ->
+         (* For every host
+          * 1. get all the pGPUs in the associated gpu_group
+          * 2. accumulate all the remaining capapcity for the vGPU
+          * 3. use the accumulated capacity as the ranking metrics*)
+         Db.Host.get_PGPUs ~__context ~self:host
+         |> Listext.List.intersect  (Db.GPU_group.get_PGPUs ~__context ~self:gpu_group)
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm_helpers.ml]: We can factor out `Db.GPU_group.get_PGPUs` and put it above the `group_by`, because the `gpu_group` doesn't change.

**Comment by robhoes** [comment]
There was a conflict after merging the "dry-run" PR due to the additional `~pre_allocate_list:[]` param that is now needed. I've fixed it.

**Pull request merged**